web/src/main/kotlin/com/example/ajaxproject/service/PrivateChatServiceImpl.kt

```diff
@@ -13,79 +13,83 @@ import org.bson.types.ObjectId
+import reactor.core.publisher.Mono
 
 @Service
 class PrivateChatServiceImpl(
     private val privateChatRoomRepository: PrivateChatRoomRepository,
     private val privateChatMessageRepository: PrivateChatMessageRepository,
-    private val userService: UserService,
+    private val userService: UserService
 ) : PrivateChatService {
 
-    companion object {
-        private val logger: Logger = LoggerFactory.getLogger(GroupChatServiceImpl::class.java)
+    override fun createPrivateRoom(senderId: String, recipientId: String): Mono<PrivateChatRoom> {
+        val roomId = roomIdFormat(senderId, recipientId)
+        return Mono.zip(
+            userService.getById(senderId),
+            userService.getById(recipientId)
+        ).flatMap {
+            val sender = it.t1
+            val recipient = it.t2
+            privateChatRoomRepository.findChatRoomById(roomId)
+                .switchIfEmpty(
+                    privateChatRoomRepository.save(
+                        PrivateChatRoom(
+                            id = roomId,
+                            senderId = sender.id,
+                            recipientId = recipient.id
+                        )
+                    )
+                )
+        }
     }
 
-    override fun createPrivateRoom(senderId: String, recipientId: String): PrivateChatRoom {
-
-        userService.getUserById(senderId)
-
-        userService.getUserById(recipientId)
-
-        val roomId = roomIdFormat(senderId, recipientId)
+    override fun getPrivateRoom(roomId: String): Mono<PrivateChatRoom> {
+        return privateChatRoomRepository.findChatRoomById(roomId)
+            .switchIfEmpty(Mono.error(NotFoundException("Room with $roomId not found")))
+    }
 
-        logger.info("Create new private room {}", roomId)
+    override fun sendPrivateMessage(privateMessageDTO: PrivateMessageDTO): Mono<PrivateChatMessage> {
+        return createPrivateRoom(privateMessageDTO.senderId, privateMessageDTO.recipientId)
+            .flatMap { room ->
+                val privateChatMessage = PrivateChatMessage(
+                    id = ObjectId(),
+                    privateChatRoomId = room.id,
+                    message = privateMessageDTO.message,
+                    senderId = privateMessageDTO.senderId
+                )
+                privateChatMessageRepository.save(privateChatMessage)
+                    .doOnSuccess { logger.info("Send private message {}", it.id) }
+            }
+    }
 
-        return privateChatRoomRepository.findChatRoomById(roomId)
-            ?: privateChatRoomRepository.save(
-                PrivateChatRoom(
-                    id = roomId,
-                    senderId = senderId,
-                    recipientId = recipientId
+    override fun getAllPrivateMessages(roomDTO: RoomDTO): Mono<List<PrivateChatMessage>> {
+        return userService.getById(roomDTO.senderId)
+            .map { userService.getById(roomDTO.recipientId) }
```

Comment:
Looks wierd.
Maybe zip it?
```kotlin
Mono.zip(
    userService.getById(roomDTO.senderId),
    userService.getById(roomDTO.recipientId),
)
```

Reply:
I think that it also do not need. Here the logic can be further simplified